// @dhis2/app-adapter/build/cjs/components/AppWrapper.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AppWrapper = void 0;
var _style = _interopRequireDefault(require("styled-jsx/style"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _react = _interopRequireDefault(require("react"));
var _useLocale = require("../utils/useLocale.js");
var _useVerifyLatestUser = require("../utils/useVerifyLatestUser.js");
var _Alerts = require("./Alerts.js");
var _ConnectedHeaderBar = require("./ConnectedHeaderBar.js");
var _ErrorBoundary = require("./ErrorBoundary.js");
var _LoadingMask = require("./LoadingMask.js");
var _AppWrapperStyle = require("./styles/AppWrapper.style.js");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const AppWrapper = _ref => {
  let {
    children,
    plugin,
    onPluginError,
    clearPluginError,
    direction: configDirection
  } = _ref;
  const {
    loading: localeLoading,
    direction: localeDirection
  } = (0, _useLocale.useCurrentUserLocale)(configDirection);
  const {
    loading: latestUserLoading
  } = (0, _useVerifyLatestUser.useVerifyLatestUser)();
  if (localeLoading || latestUserLoading) {
    return /*#__PURE__*/_react.default.createElement(_LoadingMask.LoadingMask, null);
  }
  if (plugin) {
    return /*#__PURE__*/_react.default.createElement("div", {
      className: `jsx-${_AppWrapperStyle.styles.__hash}` + " " + "app-shell-adapter"
    }, /*#__PURE__*/_react.default.createElement(_style.default, {
      id: _AppWrapperStyle.styles.__hash
    }, _AppWrapperStyle.styles), /*#__PURE__*/_react.default.createElement("div", {
      className: `jsx-${_AppWrapperStyle.styles.__hash}` + " " + "app-shell-app"
    }, /*#__PURE__*/_react.default.createElement(_ErrorBoundary.ErrorBoundary, {
      plugin: true,
      onPluginError: onPluginError,
      onRetry: () => {
        if (clearPluginError) {
          clearPluginError();
        }
        window.location.reload();
      }
    }, children)), /*#__PURE__*/_react.default.createElement(_Alerts.Alerts, null));
  }
  return /*#__PURE__*/_react.default.createElement("div", {
    className: `jsx-${_AppWrapperStyle.styles.__hash}` + " " + "app-shell-adapter"
  }, /*#__PURE__*/_react.default.createElement(_style.default, {
    id: _AppWrapperStyle.styles.__hash
  }, _AppWrapperStyle.styles), /*#__PURE__*/_react.default.createElement("div", {
    dir: localeDirection,
    className: `jsx-${_AppWrapperStyle.styles.__hash}`
  }, /*#__PURE__*/_react.default.createElement(_ConnectedHeaderBar.ConnectedHeaderBar, null)), /*#__PURE__*/_react.default.createElement("div", {
    className: `jsx-${_AppWrapperStyle.styles.__hash}` + " " + "app-shell-app"
  }, /*#__PURE__*/_react.default.createElement(_ErrorBoundary.ErrorBoundary, {
    onRetry: () => window.location.reload()
  }, children)), /*#__PURE__*/_react.default.createElement(_Alerts.Alerts, null));
};
exports.AppWrapper = AppWrapper;
AppWrapper.propTypes = {
  children: _propTypes.default.node,
  clearPluginError: _propTypes.default.func,
  direction: _propTypes.default.oneOf(['ltr', 'rtl', 'auto']),
  plugin: _propTypes.default.bool,
  onPluginError: _propTypes.default.func
};
