// @dhis2/app-adapter/build/cjs/components/ErrorBoundary.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ErrorBoundary = void 0;
var _style = _interopRequireDefault(require("styled-jsx/style"));
var _d2I18n = _interopRequireDefault(require("@dhis2/d2-i18n"));
var _classnames = _interopRequireDefault(require("classnames"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _react = _interopRequireWildcard(require("react"));
var _ButtonStyle = _interopRequireDefault(require("./styles/Button.style.js"));
var _ErrorBoundaryStyle = _interopRequireDefault(require("./styles/ErrorBoundary.style.js"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
// In order to avoid using @dhis2/ui components in the error boundary - as anything
// that breaks within it will not be caught properly - we define a component
// with the same styles as Button
const UIButton = _ref => {
  let {
    children,
    onClick,
    plugin
  } = _ref;
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_style.default, {
    id: _ButtonStyle.default.__hash
  }, _ButtonStyle.default), /*#__PURE__*/_react.default.createElement("button", {
    onClick: onClick,
    className: `jsx-${_ButtonStyle.default.__hash}` + " " + ((plugin ? 'pluginButton' : null) || "")
  }, children));
};
UIButton.propTypes = {
  children: _propTypes.default.node.isRequired,
  onClick: _propTypes.default.func.isRequired,
  plugin: _propTypes.default.bool
};
const InfoIcon24 = () => /*#__PURE__*/_react.default.createElement("svg", {
  height: "24",
  viewBox: "0 0 24 24",
  width: "24",
  xmlns: "http://www.w3.org/2000/svg"
}, /*#__PURE__*/_react.default.createElement("path", {
  d: "m12 2c5.5228475 0 10 4.4771525 10 10s-4.4771525 10-10 10-10-4.4771525-10-10 4.4771525-10 10-10zm0 2c-4.418278 0-8 3.581722-8 8s3.581722 8 8 8 8-3.581722 8-8-3.581722-8-8-8zm1 7v6h-2v-6zm-1-4c.5522847 0 1 .44771525 1 1s-.4477153 1-1 1-1-.44771525-1-1 .4477153-1 1-1z",
  fill: "#A0ADBA"
}));
const translatedErrorHeading = _d2I18n.default.t('An error occurred in the DHIS2 application.');
class ErrorBoundary extends _react.Component {
  constructor(props) {
    super(props);
    _defineProperty(this, "toggleTechInfoDrawer", () => {
      this.setState({
        drawerOpen: !this.state.drawerOpen
      });
    });
    _defineProperty(this, "handleCopyErrorDetails", () => {
      const errorDetails = this.errorDetailsRef.current.textContent;
      navigator.clipboard.writeText(errorDetails).then(() => {
        alert(_d2I18n.default.t('Technical details copied to clipboard'));
      });
    });
    _defineProperty(this, "handleCopyErrorDetailsPlugin", _ref2 => {
      let {
        error,
        errorInfo
      } = _ref2;
      const errorDetails = `${error}\n${error === null || error === void 0 ? void 0 : error.stack}\n${errorInfo === null || errorInfo === void 0 ? void 0 : errorInfo.componentStack}`;
      navigator.clipboard.writeText(errorDetails).then(() => {
        alert(_d2I18n.default.t('Technical details copied to clipboard'));
      });
    });
    _defineProperty(this, "handleSafeLoginRedirect", () => {
      window.location.href = this.props.baseURL + (this.props.baseURL.endsWith('/') ? '' : '/') + 'dhis-web-commons/security/login.action';
    });
    this.state = {
      error: null,
      errorInfo: null,
      drawerOpen: false
    };
    this.errorDetailsRef = /*#__PURE__*/_react.default.createRef();
  }
  componentDidCatch(error, errorInfo) {
    if (this.props.plugin) {
      if (this.props.onPluginError) {
        console.error(error);
        this.props.onPluginError(error);
      }
    }
    this.setState({
      error,
      errorInfo
    });
  }
  render() {
    const {
      children,
      fullscreen,
      onRetry,
      loginApp,
      baseURL
    } = this.props;
    if (this.state.error) {
      if (this.props.plugin) {
        return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_style.default, {
          id: _ErrorBoundaryStyle.default.__hash
        }, _ErrorBoundaryStyle.default), /*#__PURE__*/_react.default.createElement("div", {
          className: `jsx-${_ErrorBoundaryStyle.default.__hash}` + " " + "pluginBoundary"
        }, /*#__PURE__*/_react.default.createElement(InfoIcon24, null), /*#__PURE__*/_react.default.createElement("div", {
          className: `jsx-${_ErrorBoundaryStyle.default.__hash}` + " " + "pluginErrorMessage"
        }, _d2I18n.default.t('There was a problem loading this plugin')), /*#__PURE__*/_react.default.createElement("div", {
          onClick: () => {
            this.handleCopyErrorDetailsPlugin({
              error: this.state.error,
              errorInfo: this.state.errorInfo
            });
          },
          className: `jsx-${_ErrorBoundaryStyle.default.__hash}` + " " + "pluginErrorCopy"
        }, _d2I18n.default.t('Copy debug info to clipboard')), onRetry && /*#__PURE__*/_react.default.createElement("div", {
          className: `jsx-${_ErrorBoundaryStyle.default.__hash}` + " " + "pluginRetry"
        }, /*#__PURE__*/_react.default.createElement(UIButton, {
          onClick: onRetry,
          plugin: true
        }, _d2I18n.default.t('Try again')))));
      }
      return /*#__PURE__*/_react.default.createElement("div", {
        className: `jsx-${_ErrorBoundaryStyle.default.__hash}` + " " + ((0, _classnames.default)('mask', {
          fullscreen
        }) || "")
      }, /*#__PURE__*/_react.default.createElement(_style.default, {
        id: _ErrorBoundaryStyle.default.__hash
      }, _ErrorBoundaryStyle.default), /*#__PURE__*/_react.default.createElement("div", {
        className: `jsx-${_ErrorBoundaryStyle.default.__hash}` + " " + "container"
      }, /*#__PURE__*/_react.default.createElement("h1", {
        className: `jsx-${_ErrorBoundaryStyle.default.__hash}` + " " + "message"
      }, _d2I18n.default.t('Something went wrong')), loginApp && baseURL && /*#__PURE__*/_react.default.createElement("div", {
        className: `jsx-${_ErrorBoundaryStyle.default.__hash}` + " " + "retry"
      }, /*#__PURE__*/_react.default.createElement(UIButton, {
        onClick: this.handleSafeLoginRedirect
      }, _d2I18n.default.t('Redirect to safe login mode'))), onRetry && /*#__PURE__*/_react.default.createElement("div", {
        className: `jsx-${_ErrorBoundaryStyle.default.__hash}` + " " + "retry"
      }, /*#__PURE__*/_react.default.createElement(UIButton, {
        onClick: onRetry
      }, _d2I18n.default.t('Try again'))), /*#__PURE__*/_react.default.createElement("button", {
        onClick: this.toggleTechInfoDrawer,
        className: `jsx-${_ErrorBoundaryStyle.default.__hash}` + " " + "drawerToggle"
      }, this.state.drawerOpen ? _d2I18n.default.t('Hide technical details') : _d2I18n.default.t('Show technical details')), /*#__PURE__*/_react.default.createElement("div", {
        className: `jsx-${_ErrorBoundaryStyle.default.__hash}` + " " + ((0, _classnames.default)('drawer', {
          hidden: !this.state.drawerOpen
        }) || "")
      }, /*#__PURE__*/_react.default.createElement("div", {
        className: `jsx-${_ErrorBoundaryStyle.default.__hash}` + " " + "errorIntro"
      }, /*#__PURE__*/_react.default.createElement("p", {
        className: `jsx-${_ErrorBoundaryStyle.default.__hash}`
      }, translatedErrorHeading), /*#__PURE__*/_react.default.createElement("p", {
        className: `jsx-${_ErrorBoundaryStyle.default.__hash}`
      }, _d2I18n.default.t('The following information may be requested by technical support.')), /*#__PURE__*/_react.default.createElement(UIButton, {
        onClick: this.handleCopyErrorDetails
      }, _d2I18n.default.t('Copy technical details to clipboard'))), /*#__PURE__*/_react.default.createElement("pre", {
        ref: this.errorDetailsRef,
        className: `jsx-${_ErrorBoundaryStyle.default.__hash}` + " " + "errorDetails"
      }, `${this.state.error}\n`, this.state.error.stack, this.state.errorInfo.componentStack))));
    }
    return children;
  }
}
exports.ErrorBoundary = ErrorBoundary;
ErrorBoundary.propTypes = {
  children: _propTypes.default.node.isRequired,
  baseURL: _propTypes.default.string,
  fullscreen: _propTypes.default.bool,
  loginApp: _propTypes.default.bool,
  plugin: _propTypes.default.bool,
  onPluginError: _propTypes.default.func,
  onRetry: _propTypes.default.func
};
