// @dhis2/app-adapter/build/cjs/components/LoginAppWrapper.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LoginAppWrapper = void 0;
var _style = _interopRequireDefault(require("styled-jsx/style"));
var _appRuntime = require("@dhis2/app-runtime");
var _propTypes = _interopRequireDefault(require("prop-types"));
var _react = _interopRequireDefault(require("react"));
var _useLocale = require("../utils/useLocale.js");
var _Alerts = require("./Alerts.js");
var _ErrorBoundary = require("./ErrorBoundary.js");
var _LoadingMask = require("./LoadingMask.js");
var _AppWrapperStyle = require("./styles/AppWrapper.style.js");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const LoginAppWrapper = _ref => {
  let {
    children
  } = _ref;
  const {
    loading: localeLoading
  } = (0, _useLocale.useSystemDefaultLocale)();
  const {
    baseUrl
  } = (0, _appRuntime.useConfig)();
  // cannot check current user for a loginApp (no api/me)

  if (localeLoading) {
    return /*#__PURE__*/_react.default.createElement(_LoadingMask.LoadingMask, null);
  }
  return /*#__PURE__*/_react.default.createElement("div", {
    className: `jsx-${_AppWrapperStyle.styles.__hash}` + " " + "app-shell-adapter"
  }, /*#__PURE__*/_react.default.createElement(_style.default, {
    id: _AppWrapperStyle.styles.__hash
  }, _AppWrapperStyle.styles), /*#__PURE__*/_react.default.createElement("div", {
    className: `jsx-${_AppWrapperStyle.styles.__hash}` + " " + "app-shell-app"
  }, /*#__PURE__*/_react.default.createElement(_ErrorBoundary.ErrorBoundary, {
    onRetry: () => window.location.reload(),
    loginApp: true,
    baseURL: baseUrl
  }, children)), /*#__PURE__*/_react.default.createElement(_Alerts.Alerts, null));
};
exports.LoginAppWrapper = LoginAppWrapper;
LoginAppWrapper.propTypes = {
  children: _propTypes.default.node
};
