// @dhis2/app-runtime/build/cjs/Provider.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Provider = void 0;
var _appServiceAlerts = require("@dhis2/app-service-alerts");
var _appServiceConfig = require("@dhis2/app-service-config");
var _appServiceData = require("@dhis2/app-service-data");
var _appServiceOffline = require("@dhis2/app-service-offline");
var _react = _interopRequireDefault(require("react"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const Provider = _ref => {
  let {
    config,
    children,
    offlineInterface,
    plugin,
    parentAlertsAdd,
    showAlertsInPlugin
  } = _ref;
  return /*#__PURE__*/_react.default.createElement(_appServiceConfig.ConfigProvider, {
    config: config
  }, /*#__PURE__*/_react.default.createElement(_appServiceAlerts.AlertsProvider, {
    plugin: plugin,
    parentAlertsAdd: parentAlertsAdd,
    showAlertsInPlugin: showAlertsInPlugin
  }, /*#__PURE__*/_react.default.createElement(_appServiceData.DataProvider, null, /*#__PURE__*/_react.default.createElement(_appServiceOffline.OfflineProvider, {
    offlineInterface: offlineInterface
  }, children))));
};
exports.Provider = Provider;
Provider.displayName = 'DHIS2RuntimeProvider';
