// @dhis2-ui/select/build/cjs/single-select-field/single-select-field.e2e.stories.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.WithoutOptions = exports.WithValidationText = exports.WithLoading = exports.WithLabelAndRequiredStatus = exports.WithLabel = exports.WithHelpText = exports.WithFilterable = exports.WithClearableAndSelectedOption = void 0;
var _react = _interopRequireDefault(require("react"));
var _index = require("../index.js");
var _index2 = require("./index.js");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
var _default = exports.default = {
  title: 'SingleSelectField'
};
const WithLabel = () => /*#__PURE__*/_react.default.createElement(_index2.SingleSelectField, {
  label: "The label"
}, /*#__PURE__*/_react.default.createElement(_index.SingleSelectOption, {
  value: "1",
  label: "one"
}), /*#__PURE__*/_react.default.createElement(_index.SingleSelectOption, {
  value: "2",
  label: "two"
}), /*#__PURE__*/_react.default.createElement(_index.SingleSelectOption, {
  value: "3",
  label: "three"
}));
exports.WithLabel = WithLabel;
const WithHelpText = () => /*#__PURE__*/_react.default.createElement(_index2.SingleSelectField, {
  helpText: "The help text"
}, /*#__PURE__*/_react.default.createElement(_index.SingleSelectOption, {
  value: "1",
  label: "one"
}), /*#__PURE__*/_react.default.createElement(_index.SingleSelectOption, {
  value: "2",
  label: "two"
}), /*#__PURE__*/_react.default.createElement(_index.SingleSelectOption, {
  value: "3",
  label: "three"
}));
exports.WithHelpText = WithHelpText;
const WithValidationText = () => /*#__PURE__*/_react.default.createElement(_index2.SingleSelectField, {
  validationText: "The validation text"
}, /*#__PURE__*/_react.default.createElement(_index.SingleSelectOption, {
  value: "1",
  label: "one"
}), /*#__PURE__*/_react.default.createElement(_index.SingleSelectOption, {
  value: "2",
  label: "two"
}), /*#__PURE__*/_react.default.createElement(_index.SingleSelectOption, {
  value: "3",
  label: "three"
}));
exports.WithValidationText = WithValidationText;
const WithLabelAndRequiredStatus = () => /*#__PURE__*/_react.default.createElement(_index2.SingleSelectField, {
  label: "The label",
  required: true
}, /*#__PURE__*/_react.default.createElement(_index.SingleSelectOption, {
  value: "1",
  label: "one"
}), /*#__PURE__*/_react.default.createElement(_index.SingleSelectOption, {
  value: "2",
  label: "two"
}), /*#__PURE__*/_react.default.createElement(_index.SingleSelectOption, {
  value: "3",
  label: "three"
}));
exports.WithLabelAndRequiredStatus = WithLabelAndRequiredStatus;
const WithClearableAndSelectedOption = () => /*#__PURE__*/_react.default.createElement(_index2.SingleSelectField, {
  selected: "1",
  clearable: true
}, /*#__PURE__*/_react.default.createElement(_index.SingleSelectOption, {
  key: "1",
  value: "1",
  label: "One"
}));
exports.WithClearableAndSelectedOption = WithClearableAndSelectedOption;
const WithFilterable = () => /*#__PURE__*/_react.default.createElement(_index2.SingleSelectField, {
  filterable: true
}, /*#__PURE__*/_react.default.createElement(_index.SingleSelectOption, {
  key: "1",
  value: "1",
  label: "One"
}));
exports.WithFilterable = WithFilterable;
const WithLoading = () => /*#__PURE__*/_react.default.createElement(_index2.SingleSelectField, {
  loading: true
});
exports.WithLoading = WithLoading;
const WithoutOptions = () => /*#__PURE__*/_react.default.createElement(_index2.SingleSelectField, null);
exports.WithoutOptions = WithoutOptions;
